DigitalLearningSolutions.Web/Controllers/LoginController.cs

```diff
@@ -24,54 +33,96 @@ public IActionResult Index()
-                ModelState.AddModelError("password", "There is a password error.");
+                (adminUser, delegateUser) = loginService.VerifyUserDetailsAndGetClaims(model.Username, model.Password);
             }
-
-            if (!ModelState.IsValid)
+            catch (UserAccountNotFoundException e)
             {
+                ModelState.AddModelError("Username", e.Message);
                 return View("Index", model);
             }
-
-            var claims = new List<Claim>
+            catch (IncorrectPasswordLoginException e)
             {
-                new Claim(ClaimTypes.Email, "kevin.whittaker1@nhs.net"),
-                new Claim(CustomClaimTypes.UserCentreId, "101"),
-                new Claim(CustomClaimTypes.UserCentreManager, "True"),
-                new Claim(CustomClaimTypes.LearnCandidateId, "254480"),
-                new Claim(CustomClaimTypes.UserCentreAdmin, "True"),
-                new Claim(CustomClaimTypes.UserUserAdmin, "True"),
-                new Claim(CustomClaimTypes.UserContentCreator, "True"),
-                new Claim(CustomClaimTypes.UserAuthenticatedCm, "True"),
-                new Claim(CustomClaimTypes.UserPublishToAll, "True"),
-                new Claim(CustomClaimTypes.UserCentreReports, "True"),
-                new Claim(CustomClaimTypes.LearnUserAuthenticated, "True"),
-                new Claim(CustomClaimTypes.AdminCategoryId, "0"),
-                new Claim(CustomClaimTypes.IsSupervisor, "True"),
-                new Claim(CustomClaimTypes.IsTrainer, "False"),
-                new Claim(CustomClaimTypes.IsFrameworkDeveloper, "True"),
-                new Claim(CustomClaimTypes.LearnCandidateNumber, "KW1205"),
-                new Claim(CustomClaimTypes.UserForename, "Kevin"),
-                new Claim(CustomClaimTypes.UserSurname, "Whittaker(Developer)"),
-                new Claim(CustomClaimTypes.UserCentreName, "Test Centre NHSD"),
-                new Claim(CustomClaimTypes.UserAdminId, "1")
-            };
+                ModelState.AddModelError("Password", e.Message);
+                return View("Index", model);
+            }
+            catch (DelegateUserNotApprovedException)
+            {
+                return View("AccountNotApproved");
+            }
+
+            var claims = GetClaimsForSignIn(adminUser, delegateUser);
             var claimsIdentity = new ClaimsIdentity(claims, "Identity.Application");
             var authProperties = new AuthenticationProperties
             {
                 AllowRefresh = true,
-                IsPersistent = true,
+                IsPersistent = model.RememberMe,
                 IssuedUtc = DateTime.UtcNow
             };
             HttpContext.SignInAsync("Identity.Application", new ClaimsPrincipal(claimsIdentity), authProperties);
             return RedirectToAction("Index", "Home");
         }
+
+        private List<Claim> GetClaimsForSignIn(AdminUser? adminUser, DelegateUser? delegateUser)

```

Comment:
I wonder if it would make sense to pull this out into a service?

Reply:
Adding it into the LoginService should be easy enough to do.

Reply:
Actually, never mind. The function uses CustomClaimTypes, which is defined in Web.Helpers. so is not accessible in the data project services to avoid circular references. We could redefine it in the data project, but I think it would be more sensible to leave it in the Web project as claims aren't going to be used in any other project.

Reply:
Ah, that's fair!